web/functions/index.js

```diff
@@ -0,0 +1,885 @@
+            "userId": "${userId}",
+            "budgetId": "${budgetId}",
+            "name": "${req.body.name}",
+            "categoryId": "${categoryId}",
+            "expectedPrice": ${expectedPrice},
+            "actualPrice": ${actualPrice},
+            "startDate": "${req.body.startDate}",
+            "recurrence": "${req.body.recurrence}",
+            "billPayments": "${budgetRespectiveBills}",
+            "oneOffPayments": "${budgetRespectiveOneOffs}"
+        }`);
+
+    } catch (error) {
+        res.status(400).send(`${error.message}`)
+    }
+});
+
+//get budgets
+app.post('/GetBudgets', async (req, res) => {
+    res.set('Access-Control-Allow-Origin', '*');
+
+    try {
+
+        const userId = req.body.userId; 
+        const userRef = db.collection(userCollection).doc(`${userId}`); 
+
+         //check if the one-off already exists
+         const budgetDocs = await userRef.collection(budgetCollection).get();
+         if(!budgetDocs.empty) {
+
+            var budgets = [];
+            budgetDocs.docs.forEach(curBudget => {
+                var budget = {
+                    "name":`${curBudget.get("name")}`,
+                    "categoryId":`${curBudget.get("categoryId")}`,
+                    "expectedPrice":`${curBudget.get("expectedPrice")}`,
+                    "actualPrice":`${curBudget.get("actualPrice")}`,
+                    "startDate":`${curBudget.get("startDate")}`,
+                    "endDate":`${curBudget.get("endDate")}`,
+                    "recurrence":`${curBudget.get("recurrence")}`,
+                    "billPayments":`${curBudget.get("billPayments")}`,
+                    "oneOffPayments":`${curBudget.get("oneOffPayments")}`,
+                    "budgetId":`${curBudget.id}`
+                }
+
+                budgets.push(budget);
+            })
+
+            res.status(200).send(`{
+                "userId": "${userId}",
+                "budgets": "${JSON.stringify(budgets, null, 10)}"
+            }`);
+        }
+         else {
+            res.status(400).send("There are no existing budgets");
+         }
+
+    } catch (error) {
+        res.status(400).send(`${error.message}`)
+    }
+});
+
+//edit budget
+app.post('/EditBudget', async (req, res) => {
+    res.set('Access-Control-Allow-Origin', '*');
+
+    try {
+
+        const userId = req.body.userId; 
+        const userRef = db.collection(userCollection).doc(`${userId}`);
+        const budgetId = req.body.budgetId; 
+
+        //check if the budget already exists
+        const budgetExist = await userRef.collection(budgetCollection).doc(`${budgetId}`).get();
+        if(!budgetExist.exists){
+            res.status(400).send("This budget doesn't exist");
+        }
+
+        //get category of the one-off
+        var categoryExist = await userRef.collection(categoryCollection).where('name', '==', `${req.body.category}`).get();
+        var categoryId = "";
+        var categoryDoc = "";
```

Comment:
I don't need to pass in the category, just give all the data
I already have the frontend sorting the data by category